0006-bingings.md

```diff
@@ -0,0 +1,591 @@
+* `1` if the Ethash is valid;
+* `0` otherwise.
+
+###### Panics
+* If `block_number_ptr + 32` or `header_hash_ptr + 32` or `mix_hash_ptr + 32` or `difficulty_ptr + 32` point outside the memory or scratch buffer uses more memory than the limit with `MemoryAccessViolation`.
+
+###### Current bugs
+* `block_number` and `difficulty` are currently exposed as `u64` which are casted to `u256` which breaks Ethereum compatibility;
+* Currently, we also pass the length together with `header_hash_ptr` and `mix_hash_ptr` which is not necessary since
+we know their length.
+
+## Promises API
+
+```rust
+promise_create(account_id_len: u64,
+               account_id_ptr: u64,
+               method_name_len: u64,
+               method_name_ptr: u64,
+               arguments_len: u64,
+               arguments_ptr: u64,
+               amount_ptr: u64) -> u64
+```
+Creates a promise that will execute a method on account with given arguments and attaches the given amount.
+`amount_ptr` points to a slice of bytes representing `u128`.
+
+###### Panics
+* If `account_id_len + account_id_ptr` or `method_name_len + method_name_ptr` or `arguments_len + arguments_ptr`
+or `amount_ptr + 8` points outside the memory of the guest or host, with `MemoryAccessViolation`.
+
+###### Returns
+* Index of the new promise that uniquely identifies it within the current execution of the method.
+
+---
+
+```rust
+promise_then(promise_idx: u64,
+             account_id_len: u64,
+             account_id_ptr: u64,
+             method_name_len: u64,
+             method_name_ptr: u64,
+             arguments_len: u64,
+             arguments_ptr: u64,
+             amount_ptr: u64) -> u64            
+```
+Attaches the callback that is executed after promise pointed by `promise_idx` is complete.
+
+###### Panics
+* If `promise_idx` does not correspond to an existing promise panics with `InvalidPromiseIndex`.
+* If `account_id_len + account_id_ptr` or `method_name_len + method_name_ptr` or `arguments_len + arguments_ptr`
+or `amount_ptr + 8` points outside the memory of the guest or host, with `MemoryAccessViolation`.
+
+###### Returns
+* Index of the new promise that uniquely identifies it within the current execution of the method.
+
+---
+```rust
+promise_and(promise_id1: u64, promise_id2: u64) -> u64
+```
+Creates a new promise which completes at the same time both promises complete.
+
+###### Panics
+* If `promise_idx1` or `promise_idx2` do not correspond to existing promises panics with `InvalidPromiseIndex`.
+
+###### Returns
+* Index of the new promise that uniquely identifies it within the current execution of the method.
+
+---
+```rust
+promise_results_count() -> u64
+```
+If the current function is invoked by a callback we can access the execution results of the promises that
+caused the callback. This function returns the number of complete and incomplete callbacks.
+
+Note, we are only going to have incomplete callbacks once we have `promise_or` combinator.
```

Comment:
For promise_or we may always expect to have only one result provided (the first data receipt). But to identify which result it is, we may simulate positional args.

Reply:
The identification can be done with `promise_result`.